File: src/main/java/com/klarna/api/ordermanagement/client/ApiException.java
/*
 * Copyright 2014 Klarna AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * File containing the ApiException class
 */

package com.klarna.api.ordermanagement.client;

import com.klarna.api.ordermanagement.client.model.ErrorMessage;

import static java.lang.String.format;

/**
 * Generic API exception with the underlying http status and messages.
 */
public class ApiException extends RuntimeException {

    /**
     * The returned HTTP status code for this error.
     */
    public final int httpStatus;

    /**
     * The expected HTTP status, for example 200.
     */
    public final int expectedHttpStatus;

    /**
     * An error message received through the order management api.
     * <p/>
     * In the event of an error the message contains a {@link com.klarna.api.ordermanagement.client.model.ErrorMessage#correlation_id} identifying
     * this particular transaction in Klarna's systems. The correlation id may be requested by merchant support to
     * facilitate support inquiries.
     */
    public final ErrorMessage errorMessage;

    public ApiException(int receivedStatus, int expectedResponseStatus, ErrorMessage errorMessage) {
        super(format("Klarna API error! Received HTTP response status [%s] when expecting [%s], message=%s",
                receivedStatus, expectedResponseStatus, errorMessage));

        this.httpStatus = receivedStatus;
        this.expectedHttpStatus = expectedResponseStatus;
        this.errorMessage = errorMessage;
    }

}

File: Klarna.Rest/Klarna.Rest/Transport/ApiException.cs
#region Copyright Header
//-----------------------------------------------------------------------
// <copyright file="ApiException.cs" company="Klarna AB">
//     Copyright 2014 Klarna AB
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// </copyright>
//-----------------------------------------------------------------------
#endregion
namespace Klarna.Rest.Transport
{
    using System;
    using System.Net;
    using System.Runtime.Serialization;
    using Klarna.Rest.Models;

    /// <summary>
    /// Exception thrown for API error responses.
    /// </summary>
    [Serializable]
    public class ApiException : WebException
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="ApiException" /> class.
        /// </summary>
        /// <param name="message">The text of the error message.</param>
        /// <param name="statusCode">The HTTP status code.</param>
        /// <param name="errorMessage">The API error message model.</param>
        /// <param name="innerException">The nested exception</param>
        public ApiException(string message, HttpStatusCode statusCode, ErrorMessage errorMessage, Exception innerException) : base(message, innerException)
        {
            this.StatusCode = statusCode;
            this.ErrorMessage = errorMessage;
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="ApiException" /> class.
        /// </summary>
        /// <param name="serializationInfo">Serialization info</param>
        /// <param name="streamingContext">Streaming context</param>
        protected ApiException(SerializationInfo serializationInfo, StreamingContext streamingContext) : base(serializationInfo, streamingContext)
        {
        }

        /// <summary>
        /// Gets the API response status code.
        /// </summary>
        public HttpStatusCode StatusCode { get; internal set; }

        /// <summary>
        /// Gets the API error response model.
        /// </summary>
        public ErrorMessage ErrorMessage { get; internal set; }
    }
}